mac80211: rt2x00: some experimental improvements

DragonBluep opened the pull request:
1. Increase the sensitivity of the watchdog to avoid some unnecessary resets.
2. Some register value optimization for MT7620.
3. Some register value optimization for RT5592.
4. Fix a serious bug about EDCA parameters.

namiltd commented:
I'm testing it on mt7620/TP-link Archer C5 v4. Previously, resets happened once every few days. Time will tell if the fixes helped.

namiltd commented on package/kernel/mac80211/patches/rt2x00/622-rt2x00-always-calibrate-MT7620-when-switching-channels.patch:
```diff
@@ -0,0 +1,65 @@
+@@ -10763,7 +10778,6 @@ static void rt2800_init_rfcsr_6352(struc
+ 	rt2800_rfcsr_write_dccal(rt2x00dev, 17, 0x7C);
+ 
+ 	/* Do calibration and init PA/LNA */
```
Also remove the description, what it concerns is no longer there.

namiltd commented on package/kernel/mac80211/patches/rt2x00/620-rt2x00-increase-the-watchdog-sampling-frequency.patch:
```diff
@@ -0,0 +1,56 @@
+-	else
+-		rt2x00dev->txdma_busy = 0;
++	rt2x00dev->rxdma_busy = check_dma_busy_rx(reg_cfg, reg_int) ?
++				rt2x00dev->rxdma_busy + 1 : 0;
```
Code:
```
rt2x00dev->rxdma_busy = (rt2x00_get_field32(reg_cfg, WPDMA_GLO_CFG_TX_DMA_BUSY) &&
	rt2x00_get_field32(reg_int, INT_SOURCE_CSR_TX_COHERENT)) ?
	rt2x00dev->rxdma_busy + 1 : 0;
```
will probably be faster because without calling the function.

namiltd commented:
I haven't experienced a wifi reset in over a week. It seems the patch improved stability.

namiltd commented:
Please add the patch [mac80211: rt2x00: some experimental improvements](https://github.com/openwrt/openwrt/pull/16845/commits/29877140e1c540df3f04be74a9c3259587dc9ffd) to release 24.10.

DragonBluep commented:
@namiltd If you have time, perhaps you are interested in testing the hostapd patch. It should be helpful for the dead loop `did not acknowledge authentication response` reported in https://github.com/openwrt/openwrt/issues/10707.
https://github.com/DragonBluep/openwrt/commit/0ce4a37025637400c1d8c46747971cdc0e8b24c7

P.S. I increased watchdog timeout value to 4 seconds in the latest patch.

namiltd commented:
The patch worked for me without any problems in the previous version.

I did not experience the message "disconnected due to excessive missing ACKs" so I am not able to check the operation of the hostapd patch.

namiltd commented:
Question: What should be the effect of 622-rt2x00-rework-link-tuner-for-MT7620.patch? My router has RT6352 and I don't see any difference in operation.

DragonBluep commented:
> Question: What should be the effect of 622-rt2x00-rework-link-tuner-for-MT7620.patch? My router has RT6352 and I don't see any difference in operation.

A short summary, correct gain value and disable dynamic TX power for MT7620 because not implemented. Reduce CPU usage. Suppress warning `ieee80211 phy1: rt2800_config_alc_rt6352: Warning - RF busy while configuring ALC`.

hauke commented:
Please add a commit message to the patches you are adding.

If these patches should not get merged, please keep the PR in draft state.

DragonBluep commented:
I still have some pending work for RT5592, so mark it as draft.

namiltd commented:
Most likely on Saturday version 24.10-rc1 will be released. It would be nice if it also includes this PR.

namiltd commented:
Since this PR was published I have not experienced a single WiFi hang on the mt7620/TP-Link Archer C5 v4. Please finish and approve so others can benefit from it too.

namiltd commented:
Please finalize this PR and add it to branch 24.10.

namiltd commented:
@DragonBluep Please update PR info

kofec commented:
Not sure if this is an important but:
`[  145.374876] ieee80211 phy1: rt2800_rxiq_calibration: Warning - RXIQ calibration FAIL`

for today's build (18:00 23.12.2024) :
```
[    0.000000] Linux version 6.6.67 (MojeBuild@OnServer) (mipsel-openwrt-linux-musl-gcc (OpenWrt GCC 13.3.0 r28213-372afe10f0) 13.3.0, GNU ld (GNU Binutils) 2.42) #0 Mon Dec 23 11:40:35 2024
[    0.000000] Board has DDR2
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7620A ver:2 eco:6
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019650 (MIPS 24KEc)
[    0.000000] MIPS: machine is Xiaomi MiWiFi Mini
....
[   50.968079] ieee80211 phy1: rt2x00_set_rt: Info - RT chipset 6352, rev 0500 detected
[   50.976494] ieee80211 phy1: rt2x00_set_rf: Info - RF chipset 7620 detected
...
[  145.374876] ieee80211 phy1: rt2800_rxiq_calibration: Warning - RXIQ calibration FAIL
[  145.407164] br-lan: port 4(phy1-ap0) entered blocking state
[  145.412901] br-lan: port 4(phy1-ap0) entered disabled state
[  145.418919] rt2800_wmac 10180000.wmac phy1-ap0: entered allmulticast mode
[  145.426570] rt2800_wmac 10180000.wmac phy1-ap0: entered promiscuous mode
[  148.584921] br-lan: port 4(phy1-ap0) entered blocking state
[  148.590660] br-lan: port 4(phy1-ap0) entered forwarding state
[  150.094975] br-lan: port 5(phy1-ap1) entered blocking state
[  150.100759] br-lan: port 5(phy1-ap1) entered disabled state
[  150.106552] rt2800_wmac 10180000.wmac phy1-ap1: entered allmulticast mode
[  150.113858] rt2800_wmac 10180000.wmac phy1-ap1: entered promiscuous mode
[  150.154858] rt2800_wmac 10180000.wmac phy1-ap1: left allmulticast mode
[  150.161573] rt2800_wmac 10180000.wmac phy1-ap1: left promiscuous mode
[  150.168351] br-lan: port 5(phy1-ap1) entered disabled state
[  150.247857] br-lan: port 5(phy1-ap1) entered blocking state
[  150.253593] br-lan: port 5(phy1-ap1) entered disabled state
[  150.259441] rt2800_wmac 10180000.wmac phy1-ap1: entered allmulticast mode
[  150.266771] rt2800_wmac 10180000.wmac phy1-ap1: entered promiscuous mode
```

namiltd commented:
@kofec Did you compile with or without this patch?

kofec commented:
> @kofec Did you compile with or without this patch?

With this patch - 
wget https://patch-diff.githubusercontent.com/raw/openwrt/openwrt/pull/16845.patch
and
git apply 16845.patch

namiltd commented:
Did the problem persist without this patch?

namiltd commented on package/kernel/mac80211/patches/rt2x00/620-06-rt2x00-fix-register-operation-on-RXIQ-calibration.patch:
```diff
@@ -0,0 +1,60 @@
+ 
+ 	bbpval = bbp4 & (~0x18);
+-	bbpval = bbp4 | 0x00;
++	bbpval = bbpval | 0x00;
```
I wonder what this code is for, it does nothing.

DragonBluep replied:
Just for code review, the compiler can optimize them.

DragonBluep commented:
> @DragonBluep Please update PR info.

Done.

> 
> I would also suggest removing the word experimental from the patches. Because it would mean they were not tested.

I really don't have time for long-term testing haha~

DragonBluep commented:
> Not sure if this is an important but:
> `[ 145.374876] ieee80211 phy1: rt2800_rxiq_calibration: Warning - RXIQ calibration FAIL`

@kofec @namiltd This issue already here for a long time. The MTK proprietary driver sometimes also fail in RXIQ calibration. Sometimes we just need some luck or a cold boot.

DragonBluep commented:
> I'm not sure that point `4. Correct MT7620 clock cycle count.` is needed. Vendor driver never correct this value for SoC (builtin variant of wifi chips). Only in programming guide the meaning of `US_CYC` field is explained but exact values are shown just for USB and PCIe bus. Possiblly bug in vendor driver?

I think these built-in wifi never use this register. I don‘t see any different after applying this patch. I will do some further investigations later.

DragonBluep commented:
This PR is ready now. I'd like to backport it to the v24 branch before first LTS release.

I found another serious bug about ralink hardware queue index. I need some time to address it. I can push the fix before the end of this week. So if we don't plan to release the first v24 LTS this week, this PR can be put on hold for a while.

namiltd commented:
Tested on ramips/mt7620(RT chipset 6352, rev 0500)/TP-Linik-Archer-C5-v4

123serge123 commented:
@DragonBluep 
> I think these built-in wifi never use this register. I don‘t see any different after applying this patch. I will do some further investigations later.

Wifi MAC always use this register. You can think of it as time calibration. If US_CYC is incorrect all other times will be incorrect too. See decriptions of regs TX_SW_CGF[012], TX_TIMEOUT_CFG or XIFS_TIME_CFG for example.
And after corecttion of US_CYC I see not great by mesurable improvement in rx directions for all SoCs.

namiltd commented:
I propose to approve the current commits. The next ones will be in 24.10.1 at worst.

DragonBluep committed:
mac80211: rt2x00: some improvements for rt5592

This patch contains some performance and stability improvements
for RT5592:

1. Fix RFCSR register init values for RT5592.
2. Fix BBP register init values for RT5592.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>
Link: https://github.com/openwrt/openwrt/pull/16845
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
mac80211: rt2x00: some improvements for mt7620

This patch contains some performance and stability improvements
for MT7620:

1. Always calibrate MT7620 when switching channel.
2. Rework link tuner for MT7620.
3. Correct MT7620 SDM mode register value.
4. Fix register operation on RXIQ calibration.
5. Fix TX_PIN_CFG and TX_ALC_VGA3 init values.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>
Link: https://github.com/openwrt/openwrt/pull/16845
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
mac80211: rt2x00: some improvements for rt2800 generic

1. Respect the rt2800 hardware TX queue index.
2. Increase the watchdog sampling frequency.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>
Link: https://github.com/openwrt/openwrt/pull/16845
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep commented:
> @DragonBluep
> 
> > I think these built-in wifi never use this register. I don‘t see any different after applying this patch. I will do some further investigations later.
> 
> Wifi MAC always use this register. You can think of it as time calibration. If US_CYC is incorrect all other times will be incorrect too. See descriptions of regs TX_SW_CGF[012], TX_TIMEOUT_CFG or XIFS_TIME_CFG for example. And after corecttion of US_CYC I see not great but mesurable improvement in rx direction for all SoCs.

I did some simple tests about this register but don't see much differences. Anyway, I removed that patch. Ans I found there was a clock bug so I pushed another patch 620-01-rt2x00-fix-clock-init-for-rt2800-SoC.patch to fix it.
```
speed test with iperf3 on MT7620:

		BW20 2x2	BW20 2x2	BW40 2x2	BW40 2x2
US_CYC_CNT	lan --> wlan	wlan --> lan	lan --> wlan	wlan --> lan
0xc1 (bus clk)	73.9		68.6		92.5		91.5
0x21 (default)	75.0		61.9		94.8		86.3
```
P.S. These test results are unreliable because there are too many disturbances in 2.4GHz.

DragonBluep commented:
> This PR is ready now. I'd like to backport it to the v24 branch before first LTS release.
> 
> I found another serious bug about ralink hardware queue index. I need some time to address it. I can push the fix before the end of this week. So if we don't plan to release the first v24 LTS this week, this PR can be put on hold for a while.

The fix has been merged into patch 620-02-rt2x00-respect-rt2800-hardware-TX-queue-index.patch

DragonBluep commented on package/kernel/mac80211/patches/rt2x00/620-01-rt2x00-fix-clock-init-for-rt2800-SoC.patch:
```diff
@@ -0,0 +1,91 @@
+-			clk = clk_get_sys("cpu", NULL);
+-
+-			if (IS_ERR(clk)) {
+-				rate = 125;
```
@paraka Hi! Would you mind have a review on this patch? Is it a correct fix? I noticed my MT7620 board always use rate 125 to initialize the US_CYC_CNT.

paraka replied:
@DragonBluep : You should just get wmac clock here. That is the reason of that introduced in the driver as periph clock with xtal as parent. I don't understand why are you trying to get the bus and xtal clocks here... If the clock plan is correct (it should be) only node index for wmac in dts file must be used. No other change should be needed in dts. If there is something to fix regarding the clock it should be done at the driver side and just consume correct clock here.

123serge123 commented:
@DragonBluep 
It seams that there were changes in `clk_get_sys` semantics since kernel 6.5. In kernel 6.4 and before `clk_get_sys("cpu", NULL)` was enough to get fixed rate cpu clock. As workaround try to add `bus` as 193MHz (for mt7620 SoC) fixed rate clock in dts-file.

namiltd commented:
PR needs to be rewritten since mac80211 has been updated to 6.12.6: https://github.com/openwrt/openwrt/commit/a85059438f2caed0d01a87fce9b8bd5bed134bb2

namiltd commented:
Unfortunately today "rt2800_watchdog_dma_busy: Warning - Watchdog TX DMA busy detected" appeared twice again.
I don't know what's causing this, but the more stable version was https://github.com/openwrt/openwrt/commit/896feddf19429d76a9c59f600b84a0f537020c35

DragonBluep commented:
> @DragonBluep It seams that there were changes in `clk_get_sys` semantics since kernel 6.5. In kernel 6.4 and before `clk_get_sys("cpu", NULL)` was enough to get fixed rate cpu clock. As workaround try to add `bus` as 193MHz (for mt7620 SoC) fixed rate clock in dts-file.

MT7620 CPU and BUS clocks are adjustable, so not all devices use 193MHz (cpu_clk / 3).

DragonBluep commented:
> PR needs to be refreshed since mac80211 has been updated to 6.12.6: [a850594](https://github.com/openwrt/openwrt/commit/a85059438f2caed0d01a87fce9b8bd5bed134bb2)

Done. No new diff changes.

> Unfortunately today "rt2800_watchdog_dma_busy: Warning - Watchdog TX DMA busy detected" appeared twice again. I don't know what's causing this, but the more stable version was [896fedd](https://github.com/openwrt/openwrt/commit/896feddf19429d76a9c59f600b84a0f537020c35)

This is expected. The watchdog can recovery the strange rt2x00 DMA hang issue. If you don't suffer from it, you can use module parameter to disable it.
```
echo "rt2800lib watchdog=0" > /etc/modules.d/rt2800lib ; reboot
```

hauke commented:
> > @DragonBluep It seams that there were changes in `clk_get_sys` semantics since kernel 6.5. In kernel 6.4 and before `clk_get_sys("cpu", NULL)` was enough to get fixed rate cpu clock. As workaround try to add `bus` as 193MHz (for mt7620 SoC) fixed rate clock in dts-file.
> 
> MT7620 CPU and BUS clocks are adjustable, so not all devices use 193MHz (cpu_clk / 3).

The driver should request the clock which is forwarded the the Wifi IP block in the SoC. The Wifi IP block is connected using multiple signals (wires) with other IP blocks like the address bus, the interrupt controller, the clock generator and others. 
The wifi driver should request the clocks it is connected to. The wifi driver can also request the correct clock rate from the clk driver. 

I do not know the details of the ralink platform. I am just talking about the general architecture in Linux.

paraka commented:
> > @DragonBluep It seams that there were changes in `clk_get_sys` semantics since kernel 6.5. In kernel 6.4 and before `clk_get_sys("cpu", NULL)` was enough to get fixed rate cpu clock. As workaround try to add `bus` as 193MHz (for mt7620 SoC) fixed rate clock in dts-file.
> 
> MT7620 CPU and BUS clocks are adjustable, so not all devices use 193MHz (cpu_clk / 3).

This can be true but you have to adjust properly in the driver side. For the cpu currently we have:
- https://elixir.bootlin.com/linux/v6.13-rc2/source/drivers/clk/ralink/clk-mtmips.c#L679
An for the bus clock we are adjusting already depending on some dividers:
- https://elixir.bootlin.com/linux/v6.13-rc2/source/drivers/clk/ralink/clk-mtmips.c#L620
This is what the original arch code was doing and it should do the correct thing already here. What is wrong with this? Which extra adjustements are missing? The clock for the wmac must be transparent and have a proper rate for every single ralink SoC just requesting the clock index in the wmac device tree node. Any other change is not coherent with the rest of the device tree nodes nor a good way to go at all. So, please, fix it in the driver side or provide what is missing so I can help you to properly adjust this.
Thanks!

DragonBluep commented:
> > > @DragonBluep It seams that there were changes in `clk_get_sys` semantics since kernel 6.5. In kernel 6.4 and before `clk_get_sys("cpu", NULL)` was enough to get fixed rate cpu clock. As workaround try to add `bus` as 193MHz (for mt7620 SoC) fixed rate clock in dts-file.
> > 
> > 
> > MT7620 CPU and BUS clocks are adjustable, so not all devices use 193MHz (cpu_clk / 3).
> 
> This can be true but you have to adjust properly in the driver side. For the cpu currently we have:
> 
> * https://elixir.bootlin.com/linux/v6.13-rc2/source/drivers/clk/ralink/clk-mtmips.c#L679
> 
> And for the bus clock we are adjusting already depending on some dividers:
> 
> * https://elixir.bootlin.com/linux/v6.13-rc2/source/drivers/clk/ralink/clk-mtmips.c#L620
> 
> This is what the original arch code was doing and it should do the correct thing already here. What is wrong with this? Which extra adjustements are missing? The clock for the wmac must be transparent and have a proper rate for every single ralink SoC just requesting the clock index in the wmac device tree node. Any other change is not coherent with the rest of the device tree nodes nor a good way to go at all. So, please, fix it in the driver side or provide what is missing so I can help you to properly adjust this. Thanks!

@paraka @123serge123 Sorry for the late. I did some tests about this issue.
There are two clock source inputs in the rt2x00 wireless driver. The main clock is assigned in device tree and defined as 
[`"10180000.wmac"`](https://elixir.bootlin.com/linux/v6.6.67/source/drivers/clk/ralink/clk-mtmips.c#L210) in the new ralink clock driver. It is currently working well. Another clock source is connected to the `bus`(aka system clock), I guess it's used to provide microsecond systick for the WiFi IP internal MCU. the systick frequency is same as the `bus` clock.

we are using [`clk_get_sys("bus", NULL);`](https://elixir.bootlin.com/linux/v5.15.175/C/ident/clk_get_sys) to get the `bus` clock in rt2x00, but I found it does not work now. I think this is caused by the new introduced ralink clock driver. In the old ralink clock driver, after calling [`ralink_clk_add()`](https://elixir.bootlin.com/linux/v5.15.175/source/arch/mips/ralink/clk.c#L19), the registered clock will be added to a static list [`clocks`](https://elixir.bootlin.com/linux/v5.15.175/source/drivers/clk/clkdev.c#L24). And the `clk_get_sys()` can query the clock name from this list to get the clock. However, the new ralink clock driver won't add the registered clock to the `clocks` list, so `clk_get_sys()` always return NULL.

I'm not sure if it's the expected behavior with the new ralink clock driver. If yes, maybe we need to add another clock source in the clock driver and assign it in the device tree. Some things like this:
```
--- a/clk-mtmips.c
+++ b/clk-mtmips.c
@@ -208,7 +208,8 @@ static struct mtmips_clk mt7620_pherip_clks[] = {
 	{ CLK_PERIPH("10000b40.spi", "bus") },
 	{ CLK_PERIPH("10000c00.uartlite", "periph") },
 	{ CLK_PERIPH("10130000.mmc", "sdhc") },
-	{ CLK_PERIPH("10180000.wmac", "xtal") }
+	{ CLK_PERIPH("10180000.wmac", "xtal") },
+	{ CLK_PERIPH("10180000.sys", "bus") }
 };
```

Regards,

paraka commented:
[snip]

> 
> @paraka @123serge123 Sorry for the late. I did some tests about this issue. There are two clock source inputs in the rt2x00 wireless driver. The main clock is assigned in device tree and defined as [`"10180000.wmac"`](https://elixir.bootlin.com/linux/v6.6.67/source/drivers/clk/ralink/clk-mtmips.c#L210) in the new ralink clock driver. It is currently working well. Another clock source is connected to the `bus`(aka system clock), I guess it's used to provide microsecond systick for the WiFi IP internal MCU. the systick frequency is same as the `bus` clock.
> 

Thanks for explanation.

> we are using [`clk_get_sys("bus", NULL);`](https://elixir.bootlin.com/linux/v5.15.175/C/ident/clk_get_sys) to get the `bus` clock in rt2x00, but I found it does not work now. I think this is caused by the new introduced ralink clock driver. In the old ralink clock driver, after calling [`ralink_clk_add()`](https://elixir.bootlin.com/linux/v5.15.175/source/arch/mips/ralink/clk.c#L19), the registered clock will be added to a static list [`clocks`](https://elixir.bootlin.com/linux/v5.15.175/source/drivers/clk/clkdev.c#L24). And the `clk_get_sys()` can query the clock name from this list to get the clock. However, the new ralink clock driver won't add the registered clock to the `clocks` list, so `clk_get_sys()` always return NULL.

Does this happen from the beggining? Since when this is broken? I am not aware of any semantics changes in clk_get_sys() like @123serge123 is saying? Can you point me out to a link about this change? Does using con_id instead of dev_id in clk_get_sys also return NULL? I mean:

```
clk = clk_get_sys(NULL, "bus");
```

> 
> I'm not sure if it's the expected behavior with the new ralink clock driver. If yes, maybe we need to add another clock source in the clock driver and assign it in the device tree. Some things like this:
> 
> ```
> --- a/clk-mtmips.c
> +++ b/clk-mtmips.c
> @@ -208,7 +208,8 @@ static struct mtmips_clk mt7620_pherip_clks[] = {
>  	{ CLK_PERIPH("10000b40.spi", "bus") },
>  	{ CLK_PERIPH("10000c00.uartlite", "periph") },
>  	{ CLK_PERIPH("10130000.mmc", "sdhc") },
> -	{ CLK_PERIPH("10180000.wmac", "xtal") }
> +	{ CLK_PERIPH("10180000.wmac", "xtal") },
> +	{ CLK_PERIPH("10180000.sys", "bus") }
>  };
> ```

This would be a preferred solution that using fixed clocks in dts, yes. But I think we should clarify first the clk_get_sys() current behaviour.

Thanks,
    Sergio Paracuellos
> Regards,

DragonBluep commented:
> [snip]
> 
> > @paraka @123serge123 Sorry for the late. I did some tests about this issue. There are two clock source inputs in the rt2x00 wireless driver. The main clock is assigned in device tree and defined as [`"10180000.wmac"`](https://elixir.bootlin.com/linux/v6.6.67/source/drivers/clk/ralink/clk-mtmips.c#L210) in the new ralink clock driver. It is currently working well. Another clock source is connected to the `bus`(aka system clock), I guess it's used to provide microsecond systick for the WiFi IP internal MCU. the systick frequency is same as the `bus` clock.
> 
> Thanks for explanation.
> 
> > we are using [`clk_get_sys("bus", NULL);`](https://elixir.bootlin.com/linux/v5.15.175/C/ident/clk_get_sys) to get the `bus` clock in rt2x00, but I found it does not work now. I think this is caused by the new introduced ralink clock driver. In the old ralink clock driver, after calling [`ralink_clk_add()`](https://elixir.bootlin.com/linux/v5.15.175/source/arch/mips/ralink/clk.c#L19), the registered clock will be added to a static list [`clocks`](https://elixir.bootlin.com/linux/v5.15.175/source/drivers/clk/clkdev.c#L24). And the `clk_get_sys()` can query the clock name from this list to get the clock. However, the new ralink clock driver won't add the registered clock to the `clocks` list, so `clk_get_sys()` always return NULL.
> 
> Does this happen from the beggining? Since when this is broken? I am not aware of any semantics changes in clk_get_sys() like @123serge123 is saying? Can you point me out to a link about this change? 

I'm not sure. I haven't tracked it. I'll see if the older kernel and ralink clock driver behave the same later.

>Does using con_id instead of dev_id in clk_get_sys also return NULL? I mean:
> 
> ```
> clk = clk_get_sys(NULL, "bus");
> ```
> 

Not work. I've already tried it. The `clk_get_sys()` called the [clk_find()](https://elixir.bootlin.com/linux/v5.15.175/source/drivers/clk/clkdev.c#L36). However, this [clocks](https://elixir.bootlin.com/linux/v5.15.175/source/drivers/clk/clkdev.c#L24) list seems empty, So the iterator [list_for_each_entry](https://elixir.bootlin.com/linux/v5.15.175/source/drivers/clk/clkdev.c#L48) won't run at all.


> > I'm not sure if it's the expected behavior with the new ralink clock driver. If yes, maybe we need to add another clock source in the clock driver and assign it in the device tree. Some things like this:
> > ```
> > --- a/clk-mtmips.c
> > +++ b/clk-mtmips.c
> > @@ -208,7 +208,8 @@ static struct mtmips_clk mt7620_pherip_clks[] = {
> >  	{ CLK_PERIPH("10000b40.spi", "bus") },
> >  	{ CLK_PERIPH("10000c00.uartlite", "periph") },
> >  	{ CLK_PERIPH("10130000.mmc", "sdhc") },
> > -	{ CLK_PERIPH("10180000.wmac", "xtal") }
> > +	{ CLK_PERIPH("10180000.wmac", "xtal") },
> > +	{ CLK_PERIPH("10180000.sys", "bus") }
> >  };
> > ```
> 
> This would be a preferred solution that using fixed clocks in dts, yes. But I think we should clarify first the clk_get_sys() current behaviour.
> 
> Thanks, Sergio Paracuellos
> 
> > Regards,

hauke commented:
Did you send the patches upstream too?

DragonBluep commented:
> Is the change to the clocks essential for this PR? If the user facing bugs are fixed with the other changes, I would prefer if you move the clock changes into a different PR, then I can merge the other stuff.

I have removed the patches regarding the clock bug. We need more time for investigation and discussion. I think it's not an emergency issue. But for sure I will fix it later. This PR is ready now.

> Did you send the patches upstream too?

Not yet. I'd like to get some feedback from OpenWrt users and then send them upstream. And my Outlook email provider no longer allows basic git-email SMTP authentication. I need to take some time to see how to solve it. Maybe I need to change my email again, it's too bad.

hauke commented:
Should this also go into openwrt 24.10?

DragonBluep commented:
> Should this also go into openwrt 24.10?

Yes. I saw you have picked them to v24.10 branch. Thank you!

The upstream rt2x00 has not significant changes for a long time, so they are fully compatible with V24.10.
